### DISCUSSION TITLE
v0.2.0

### DISCUSSION THREAD

**Discussion by Inori** [Inori/GPCS4#128] [Announcements]
1. Support simple compute shader, compute ring.
2. Support cube map texture.
3. Improve GCN compiler.

<hr /><em>This discussion was created from the release <a href='https://github.com/Inori/GPCS4/releases/tag/v0.2.0'>v0.2.0</a>.</em>

**Comment by LifeIsStrange** [comment]
@Inori Sorry to spam it again but it seems like you are attempting the very ambitious task of making a GCN compiler from scratch, I believe there are existing compilers or at least GCN tooling that could help you in the goal of supporting GCN.
Firslty, there is the LLVM [AMDGPU ](https://llvm.org/docs/AMDGPUUsage.html) backend.
Then there are the **official GCN tools** from AMD https://rocmdocs.amd.com/en/latest/Programming_Guides/gcn-assembler.html
including the SP3 backend as an AMDGPU alternative
Then there is this guy that wrote a **blog** explaining the GCN compiler he made himself: https://www.codeproject.com/Articles/872477/Assembler-for-AMD-s-GCN-GPU
and there are at least three others:
> [cmingcnasm](https://code.google.com/p/cmingcnasm/) - cmingcnasm is a GCN minimal assembler written in c for GCC on a Linux platform. It was created by Sylvain Bertrand. I have not played with this library but I would imagine that is fast since it is written in c. Links:  [GitHub](https://github.com/sylware/cmingcnasm),  [GoogleCode](http://code.google.com/p/cminAsm4GCN/)
[gcnasm](https://github.com/balidani/gcnasm) - Created by Daniel Bali, is an open source GCN assembler that is written in c. The code is written efficiently using low-level c so it is lighting fast. It uses the GCC compiler in Linux but can easily be modified to work in a windows environment because of its standard c includes. Many of the ideas and code references on how to build Asm4GCN came from gcnasm. Links:  [GitHub](https://github.com/balidani/GcnAsm),  [OpenWall](http://openwall.info/wiki/john/development/GCN-ISA)
[HetPas Assembler](http://realhet.wordpress.com/) (windows)- Written by Realhet, HetPas has been the only directly runnable GCN assembler for windows up to this point. I have used HetPas’s assembler for a couple of years now. It’s good for creating binary kernels or just playing around with GCN Assembly. The program implements a full and feature rich GCN Assembler using a Pascal like language for host code. It does require "Data Execution Prevention" to be turned off in windows and there are instructions on how to do this. HetPas has been updated several times and is getting more and more feature rich with each release. Recently variables were added as well.

I hope this might help you.

**Reply by Inori** [reply]
All of the compilers you point out is to compile something into GCN，but what we need is exactly the inverse: compile GCN to something else. There is no one as far as I know，except for other ps4 emulator projects.

**Reply by LifeIsStrange** [reply]
> All of the compilers you point out is to compile something into GCN，but what we need is exactly the inverse: compile GCN to something else. There is no one as far as I know，except for other ps4 emulator projects.

Yes that's right except when targeting only AMD gpus.

**Reply by Inori** [reply]
> Yes that's right except when targeting only AMD gpus.

Even if you're targeting only AMD GPU, we still can not use the GCN code directly, as far as I know there's no public interface to send GCN machine code to gpu driver except for some compute kernels.

**Reply by LifeIsStrange** [reply]
Hmm OK, what are you translating GCN to?
GCN to SPIR-V?
GCN to AMDGPU intrinsics?
GCN to Vulkan?
Orbital was targetting SPIR-V https://twitter.com/kuasha420/status/1108065845186428929

**Reply by Inori** [reply]
> what are you translating GCN to? 

Nearly all ps4 emulators are targeting spir-v

**Reply by LifeIsStrange** [reply]
Note that an atypical alternative would be to transpile GCN to [DXIL](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst) and then use a regular DXIL to SPIR-V translator although I have no clue wether this mapping would be easier for you or not.
It might or not be higher level and might or not have less semantic gap/mismatch with GCN

**Reply by Inori** [reply]
> Note that an atypical alternative would be to transpile GCN to DXIL.

That is a detour. Both DXIL and Spir-V are high level language, I see no advantage translating to DXIL first.
